### ISSUE TITLE
Unable to compile basic example for NodeMCU-32s

### ISSUE THREAD

**Issue by securelpb** [khoih-prog/ESPAsync_WiFiManager_Lite#28]
### Describe the bug

Using PlatformIO with Clion I am unable to compile the basic example in ESPAsync_WiFi due to a dependency on https://registry.platformio.org/libraries/devyte/ESPAsyncDNSServer. Per the specifications of this library, it only supports the platform espressif. 

I am able to get the example to compile successfully for the espressif platform. (Note: I had to figure out that I needed to add
**build_flags = -D USE_LITTLEFS=true** to my [env] section. This would be nice to have in the release notes)
When I compile for the espressif32 platform I get the following: (I've omitted the include path because its a few hundred lines :-) )

```
`xtensa-esp32-elf-g++ -o .pio/build/nodemcu-32s/lib1b5/AsyncTCP/AsyncTCP.cpp.o -c -Wno-frame-address -std=gnu++11 -fexceptions -fno-rtti -mlongcalls -ffunction-sections -fdata-sections -Wno-error=unused-function -Wno-error=unused-variable -Wno-error=deprecated-declarations -Wno-unused-parameter -Wno-sign-compare -ggdb -freorder-blocks -Wwrite-strings -fstack-protector -fstrict-volatile-bitfields -Wno-error=unused-but-set-variable -fno-jump-tables -fno-tree-switch-conversion -MMD -Og -g2 -ggdb2 -DPLATFORMIO=60105 -DARDUINO_NodeMCU_32S -DUSE_LITTLEFS=true -DHAVE_CONFIG_H -DMBEDTLS_CONFIG_FILE=\"mbedtls/esp_config.h\" -DUNITY_INCLUDE_CONFIG_H -DWITH_POSIX -D_GNU_SOURCE -DIDF_VER=\"v4.4.3\" -DESP_PLATFORM -D_POSIX_READER_WRITER_LOCKS -DARDUINO_ARCH_ESP32 -DESP32 -DF_CPU=240000000L -DARDUINO=10812 -DARDUINO_VARIANT=\"nodemcu-32s\" -DARDUINO_BOARD=\"NodeMCU-32S\" -DARDUINO_PARTITION_default -D__PLATFORMIO_BUILD_DEBUG__ .pio/libdeps/nodemcu-32s/AsyncTCP/src/AsyncTCP.cpp
In file included from src/defines.h:135,
                 from src/main.cpp:7:
.pio/libdeps/nodemcu-32s/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:142:8: warning: #warning Using ESP32 Core 1.0.6 or 2.0.0+ [-Wcpp]
       #warning Using ESP32 Core 1.0.6 or 2.0.0+
        ^~~~~~~
In file included from .pio/libdeps/nodemcu-32s/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:182,
                 from src/defines.h:135,
                 from src/main.cpp:7:
.pio/libdeps/nodemcu-32s/ESPAsyncDNSServer/src/ESPAsyncDNSServer.h:4:10: fatal error: ESPAsyncUDP.h: No such file or directory

*********************************************************************
* Looking for ESPAsyncUDP.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:ESPAsyncUDP.h"
* Web  > https://registry.platformio.org/search?q=header:ESPAsyncUDP.h
*
*********************************************************************`

```

### Steps to Reproduce

With the following platform.ini, copy the contents of https://registry.platformio.org/libraries/khoih-prog/ESPAsync_WiFiManager_Lite/examples/ESPAsync_WiFi into the src directory of your PIO project. 
Rename main.ino to main.cpp and add #include <Arduino.h>
Run the platformio init for clion and get your targets fixed
Try to build for nodemcuv2:Success
Try to build for nodemcu-32s:Failure

```
[env]
build_flags = -D USE_LITTLEFS=true


[env:nodemcuv2]
monitor_speed = 57600
upload_speed = 115200
platform = espressif8266
board = nodemcuv2
framework = arduino
lib_deps=
    khoih-prog/ESPAsync_WiFiManager_Lite

[env:nodemcu-32s]
monitor_speed = 115200
platform = espressif32
board = nodemcu-32s
framework = arduino
lib_deps=
    khoih-prog/ESPAsync_WiFiManager_Lite
```

### Expected behavior

After some research, I found this old post about a similar issue: https://github.com/forkineye/ESPAsyncE131/issues/4#issuecomment-454037981
That comment said "An ESP32 build should reference AsnycUDP and not ESPAsyncUDP." This seems to line up with the fact that ESPAsyncDNSServer only supports espressif. 

I see that you have some alternate implementations of  ESPAsyncDNSServer, but the library dependencies don't reference them, so I figure you have your reasons. I did download your version of AsyncDNSServer_ESP32_W5500 and I was able to get the example to compile, but I actually don't know which version of that library I should be using for my ESP32-WROOM-32D powered NodeMCU-32s.

### Actual behavior

In file included from .pio/libdeps/nodemcu-32s/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:182,
                 from src/defines.h:135,
                 from src/main.cpp:7:
.pio/libdeps/nodemcu-32s/ESPAsyncDNSServer/src/ESPAsyncDNSServer.h:4:10: fatal error: ESPAsyncUDP.h: No such file or directory

---

### Information

Please ensure to specify the following:

* PlatformIO 6.1.5
* Ubuntu 22.04.1 lts
* CLion 2022.3.1
*  [NodeMCUv2](https://a.co/d/8TmHMvv) for ESP and [NodeMCU-32s](https://a.co/d/j8WmFlL) for ESP32

**Comment by khoih-prog** [comment]
HI @securelpb 

You need at least to read the [**Prerequisites**](https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite#prerequisites), then start with simpler Arduino IDE if you're still new with PIO, which requires more knowledge and skills.

Try to compile examples of [**ESPAsyncDNSServer**](https://github.com/devyte/ESPAsyncDNSServer) for ESP32 to see if you have `me-no-dev/AsyncTCP@>=1.1.1` installed

You can just copy an single exampple, the compile easily with PIO.

Why don't you use PIO to install this library and use its [`platformio.ini`](https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite/blob/main/platformio/platformio.ini)

https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite/blob/e925a1d1a6e448e94f126b17896cf4178ff8c7b4/platformio/platformio.ini#L41-L61

**Comment by securelpb** [comment]
Hello @khoih-prog 
First, thank you for your prompt reply. I really do appreciate that. 
I have been using platformio for some time and developing in C/C++ since the 90's. I'm not a noob. I did give your platform.ini a type and I get the following:

```
Compiling .pio/build/nodemcu-32s/lib1b5/AsyncTCP/AsyncTCP.cpp.o
In file included from src/defines.h:135,
                 from src/main.cpp:7:
.pio/libdeps/nodemcu-32s/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:142:8: warning: #warning Using ESP32 Core 1.0.6 or 2.0.0+ [-Wcpp]
       #warning Using ESP32 Core 1.0.6 or 2.0.0+
        ^~~~~~~
In file included from .pio/libdeps/nodemcu-32s/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:182,
                 from src/defines.h:135,
                 from src/main.cpp:7:
.pio/libdeps/nodemcu-32s/ESPAsyncDNSServer/src/ESPAsyncDNSServer.h:4:10: fatal error: ESPAsyncUDP.h: No such file or directory
```
Which I would attribute to the "platforms" value only mentioning **espressif** in the library.json for [ESPAsyncDNSServer](https://github.com/devyte/ESPAsyncDNSServer)

```
  "version": "1.0.0",
  "license": "LGPL-3.0",
  "frameworks": "arduino",
  "platforms":"espressif",
```
The examples for ESPAsyncDNSServer.ino also start with the following lines, which I don't need to compile to see how that will go with ESP32:
```
#include <ESP8266WiFi.h>
#include <ESPAsyncDNSServer.h>
#include <ESP8266WebServer.h>
```
In your library.json you call out the following, which is not aligned with the stated support of devte/ESPAsyncDNSServer:
```
    {
      "owner": "devyte",
      "name": "ESPAsyncDNSServer",
      "version": "^1.0.0",
      "platforms": ["espressif8266", "espressif32"]
    },
```
So I deleted your dependency devte/ESPAsyncDNSServer in from my local copy of nodemcu-32s/ESPAsync_WiFiManager_Lite/library.json and added a reference to you github repo https://github.com/khoih-prog/ESPAsyncDNSServer into my lib-deps. That actually worked!! 

So thank you for the tip, but there are two problems.
1) You don't seem to have https://github.com/khoih-prog/ESPAsyncDNSServer published to the pio registry (or at least I couldn't find it)
2) Your libray.json should contain the following once you publish your library:
```
    {
      "owner": "devyte",
      "name": "ESPAsyncDNSServer",
      "version": "^1.0.0",
      "platforms": ["espressif8266"]
    },
    {
      "owner": "khoih-prog",
      "name": "ESPAsyncDNSServer",
      "version": "^1.0.0",
      "platforms": [ "espressif32"]
    },
```
For the record, I was able to compile your ESPAsync_WiFi project for ESP32 with these lib_deps:

```
[env:nodemcu-32s]
monitor_speed = 115200
platform = espressif32
board = nodemcu-32s
framework = arduino
lib_deps =
 ; PlatformIO 4.x
 ;  ESP Async WebServer@>=1.2.3
 ;  https://github.com/khoih-prog/ESPAsyncWebServer
 ;  ESPAsyncTCP@>=1.2.2
 ;  https://github.com/khoih-prog/ESPAsyncTCP
 ;  AsyncTCP@>=1.1.1
 ;  ESP_DoubleResetDetector@>=1.3.2
 ;  ESP_MultiResetDetector@>=1.3.2
 ;  ESPAsyncDNSServer@>=1.0.0
 ;  LittleFS_esp32@>=1.0.6
 ; PlatformIO 5.x
 ; me-no-dev/ESP Async WebServer@>=1.2.3
  khoih-prog/ESPAsync_WiFiManager_Lite
  https://github.com/khoih-prog/ESPAsyncWebServer
  https://github.com/khoih-prog/ESPAsyncDNSServer
 ; me-no-dev/ESPAsyncTCP@>=1.2.2
  https://github.com/khoih-prog/ESPAsyncTCP
  me-no-dev/AsyncTCP@>=1.1.1
  khoih-prog/ESP_DoubleResetDetector@>=1.3.2
  khoih-prog/ESP_MultiResetDetector@>=1.3.2
 ; lorol/LittleFS_esp32@>=1.0.6
```

**Comment by khoih-prog** [comment]
> I'm not a noob

Sure you're not as you figure out so quickly how to fix the issue from a small hint.

> You don't seem to have https://github.com/khoih-prog/ESPAsyncDNSServer published to the pio registry (or at least I couldn't find it)

No, I can't do that because that is the library of `devyte` and has been long published on PIO as [**ESPAsyncDNSServer**](https://registry.platformio.org/libraries/devyte/ESPAsyncDNSServer). But it has not been well maintained to update for PIO. You can use it instead of `https://github.com/khoih-prog/ESPAsyncDNSServer` after modifying its `platformio.ini`

> Your libray.json should contain the following once you publish your library

That's the better way. But I'm still waiting the [**ESPAsyncDNSServer**](https://registry.platformio.org/libraries/devyte/ESPAsyncDNSServer) `platformio.ini` to be fixed (even a little bit **longer time than I expected**). As you're an experienced PIO user, I believe you can figure out and modify anything manually as PIO users always have to do and master.

**Comment by khoih-prog** [comment]
You can also add to `library.json` something like this, to use my forked copy, while waiting for the fix

```
{
  "name": "ESPAsyncDNSServer",
  "version": "https://github.com/khoih-prog/ESPAsyncDNSServer"
}
```

I don't know if the issue is widespread or not. If so bad, do you think I have to fix `library.json` and `platformio.ini` ?

**Comment by securelpb** [comment]
So [ESPAsyncDNSServer](https://registry.platformio.org/libraries/devyte/ESPAsyncDNSServer) isn't maintaining... That is rough. 
Thank you for showing me how to get this going. I just went back and changed only your library.json in my local copy to this and it worked. This seems like the only change you need. I actually removed the direct reference to ESPAsyncDNSServer from my platform.ini, and just let PIO unravel the dependencies. Maybe I'm lazy, but that's how I do it. I try to put the minimum in my platformio.ini and let the dependency resolver find the correct libraries. In my mind, I figure this is the best way to get what the authors have tested with, and pull in updates. 

```
    {
      "owner": "devyte",
      "name": "ESPAsyncDNSServer",
      "version": "^1.0.0",
      "platforms": ["espressif8266"]
    },
    {
      "name": "ESPAsyncDNSServer",
      "version": "https://github.com/khoih-prog/ESPAsyncDNSServer",
      "platforms": ["espressif32"]
    },
```

**Comment by khoih-prog** [comment]
Can you try with

```
{
  "name": "ESPAsyncDNSServer",
  "version": "https://github.com/khoih-prog/ESPAsyncDNSServer",
  "platforms": ["espressif8266", "espressif32"]
}
```

so we don't need 2 copies of the same library.

It seems we can't wait any more time as the issue has been wider than expected, I'll update this library and many more using the same `ESPAsyncDNSServer` after your test to confirm the change in `library.json`

**Comment by khoih-prog** [comment]
Please check the recent mods in main branch

https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite/blob/85f9435c464dde541c4dc1144cbb340588ca18e3/library.json#L25-L65


https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite/blob/85f9435c464dde541c4dc1144cbb340588ca18e3/platformio/platformio.ini#L41-L63

If they are OK, I'll create new release.

**Comment by securelpb** [comment]
Built perfectly for both espressif8266 and espressif32! Thank you so much!
I tested with your platformio.ini as well as this version, which pulls all
dependencies down from ESPAsync_WiFiManager_Lite.

[env]
build_flags = -D USE_LITTLEFS=true
lib_deps=
    https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite


[env:nodemcuv2]
monitor_speed = 57600
upload_speed = 115200
platform = espressif8266
board = nodemcuv2
framework = arduino

[env:nodemcu-32s]
monitor_speed = 115200
platform = espressif32
board = nodemcu-32s
framework = arduino

**Comment by hmueller01** [comment]
I think there is no need to define `build_flags = -D USE_LITTLEFS=true` in platformio.ini
This is set in your `defines.h` (from examples):
```
// LittleFS has higher priority than SPIFFS
#if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )
  #define USE_LITTLEFS    true
  #define USE_SPIFFS      false
#elif defined(ARDUINO_ESP32C3_DEV)
  // For core v1.0.6-, ESP32-C3 only supporting SPIFFS and EEPROM. To use v2.0.0+ for LittleFS
  #define USE_LITTLEFS          false
  #define USE_SPIFFS            true
#endif
```

Maybe it does not work as ESP_ARDUINO_VERSION_MAJOR might not be defined in PIO. I havn't checked this, as I am using just
```
// use LittleFS or SPIFFS, prefere LittleFS ...
#define USE_LITTLEFS true
#define USE_SPIFFS false
```

**Comment by securelpb** [comment]
Without that define in the build flags I get the following compile error on
espressif8266. Without that flag espressif32 does compile.

Compiling .pio/build/nodemcuv2/lib351/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
In file included from src/defines.h:135,
                 from src/main.cpp:7:
.pio/libdeps/nodemcuv2/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:
In member function 'bool ESPAsync_WiFiManager_Lite::extLoadDynamicData()':
.pio/libdeps/nodemcuv2/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:1306:12:
error: 'FileFS' was not declared in this scope
 1306 |       if (!FileFS.begin())
      |            ^~~~~~
.pio/libdeps/nodemcuv2/ESPAsync_WiFiManager_Lite/src/ESPAsync_WiFiManager_Lite.h:1316:14:
error: 'loadDynamicData' was not declared in this scope; did you mean
'hadDynamicData'?
 1316 |       return loadDynamicData();
      |              ^~~~~~~~~~~~~~~
      |              hadDynamicData
Compiling .pio/build/nodemcuv2/lib351/ESP8266WiFi/ESP8266WiFiScan.cpp.o

**Comment by khoih-prog** [comment]
Hi @securelpb @hmueller01 

IMO, you've touched the limitation of PIO, which didn't parse / take the definitions in a project `*.h` file to use throughout the dependent project files.

The current quick-fix is to manually use 

```
[env]
build_flags = -D USE_LITTLEFS=true
```

like @securelpb suggests.

The Arduino IDE compiler takes care of this scenario perfectly. I remember (not sure, no time to check) there is similar issue posted on PIO some time ago about handling complex multi-platform libraries.

You can dig-in and post this issue on [**PIO issues**](https://github.com/platformio/platformio-core/issues), with [**MRE**](https://stackoverflow.com/help/minimal-reproducible-example)

**Comment by hmueller01** [comment]
Do you use the default example `defines.h`file, or did you modify it?
Can you post, what is defined there in your config?
It looks like USE_LITTLEFS und USE_SPIFFS are getting not set to true, so the fallback to EEPROM is used:
```
  //default to use EEPROM, otherwise, use LittleFS or SPIFFS
  #if ( USE_LITTLEFS || USE_SPIFFS )

    #if USE_LITTLEFS
      #define FileFS        LittleFS
      #define FS_Name       "LittleFS"
      #warning Using LittleFS in ESPAsync_WiFiManager_Lite.h
    #else
      #define FileFS        SPIFFS
      #define FS_Name       "SPIFFS"
      #warning Using SPIFFS in ESPAsync_WiFiManager_Lite.h
    #endif

    #include <FS.h>
    #include <LittleFS.h>
  #else
    #include <EEPROM.h>
    #define FS_Name         "EEPROM"
    #define EEPROM_SIZE     2048
    #warning Using EEPROM in ESPAsync_WiFiManager_Lite.h
  #endif
```
In case of EEPROM `FileFS` gets not defined. And I forgot that case in the new `extLoadDynamicData()`.
Have to fix this. Thanks for finding this!

@khoih-prog In my first PR #20 was a much better solution, which would not have this implication!

**Comment by hmueller01** [comment]
Just a quick test. As I thought. PIO does not define `ESP_ARDUINO_VERSION_MAJOR`:
```
src/rhts-arduino.cpp:990:18: error: 'ESP_ARDUINO_VERSION_MAJOR' was not declared in this scope; did you mean 'ARDUINOJSON_VERSION_MAJOR'?
  990 |   Serial.println(ESP_ARDUINO_VERSION_MAJOR);
```
So the example will always use EEPROM on espressif8266 ...
Nevertheless I have to fix the code. It might take some time. Please be patient.

**Comment by hmueller01** [comment]
I can fully reproduce the problem. Can you please try #29 without `build_flags = -D USE_LITTLEFS=true` (using EEPROM as storage).

**Comment by khoih-prog** [comment]
Hi @securelpb 

The new [**ESPAsync_WiFiManager_Lite releases v1.10.3**](https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite/releases/tag/v1.10.3) has just been published. Your contribution is noted in [**Contributions and Thanks**](https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite#contributions-and-thanks)

Best Regards,

---

### Release v1.10.3

1. Fix compiler error if EEPROM is used. Check [fix compiler error if EEPROM is used #29](https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite/pull/29)
2. Fix compiler error in PIO. Check [Unable to compile basic example for NodeMCU-32s #28](https://github.com/khoih-prog/ESPAsync_WiFiManager_Lite/issues/28)

**Comment by securelpb** [comment]
Sorry, I’ve been tied up. Yes, the build flag is no longer required. Thank you both!
